Chore fix type issues

mdrocco opened the pull request:
This PR provides a first batch of type fixes (see #929, "pyright" section). I will follow up with a further issue about the outstanding issues that I could not figure put how to fix without a better understanding of the current behavior.

THS-on commented on keylime/ima/ima.py:
```diff
@@ -219,6 +219,7 @@ def _validate_ima_buf(
             failure = _validate_ima_ng(exclude_regex, allowlist, digest, path, hash_types="keyrings")
             if not failure:
                 # Add the key only now that it's validated (no failure)
+                assert ima_keyrings is not None
```
Under normal conditions `ima_keyrings` is never None when an IMA validator is called, but for now I would rather check it explicitly instead of an assertion.

We definitely need to refactor the agent state at some point into a single object.

mdrocco replied:
https://github.com/keylime/keylime/pull/1139/commits/e6b2344b0d1ab9b4c0d6ccfbb8399e2fe5dccf5c

THS-on commented on keylime/ima/ima_dm.py:
```diff
@@ -12,14 +12,19 @@
 from keylime.ima import ast
 from keylime.ima.dm_grammar import DM_GRAMMAR
 
+if sys.version_info >= (3, 10):
```
Maybe add a comment why this is needed for Python 3.10?

aplanas replied:
Yes, that is funny. Does lark namespace for a fixed version changes based on the Python version? I never saw that : )

mdrocco replied:
yes, at least that is what I inferred looking at the issues reported by the type checker and the code of the Lark module itself at different versions. Of course I might be wrong, feel free to double-check. I added a minimal comment for now: d84fa2c

THS-on replied:
Can you investigate with which lark versions this happens?

mdrocco replied:
My setup is Python 3.10.6 and Lark 1.1.3. If i Try the "old-style" import I get this for `Lark`:

```console
"Lark" is not exported from module "lark"
  Import from "lark.lark" instead
```

(same thing for `Transformer` and `v_args`)

THS-on replied:
Can you check if Lark 0.X versions and other Python versions, because making it conditionally on the Python version still does not seem to be right.

mdrocco replied:
Apparently it is a Lark issue and they are working on it: https://github.com/lark-parser/lark/issues/1199. To my understanding, the "old-style" import is intended to be supported, so I would simply roll back to it.

mdrocco replied:
done, hopefully the type checker will eventually be happy once they fix.

THS-on commented on keylime/ima/ima_dm.py:
```diff
@@ -293,7 +299,7 @@ def validate_table_load(self, event: "LoadEvent", match_key, digest) -> Failure:
         # Note that we get actually could get multiple lines, but this does not happen for our use cases so we
         # treat it as a failure.
-        if event.device_metadata.num_targets != len(event.targets) or used_policy["table_load"][entry] != len(
+        if event.device_metadata.num_targets != len(event.targets) or used_policy["table_load"]["num_targets"] != len(
```
That was definitely a bug. Good catch!

THS-on commented on keylime/ima/ima_dm.py:
```diff
@@ -395,6 +401,8 @@ def _check_attr(attr: Optional[Union[int, str, bool]], reference_value: Optional
     if attr is None and reference_value is not None:
         return False
 
+    assert attr is not None
```
We can just drop the `and reference_value is not None` from line 401 right? Because that is we already returned if the value was None.

mdrocco replied:
good point: 07ef85c

THS-on commented on keylime/registrar_common.py:
```diff
@@ -281,11 +281,16 @@ def do_POST(self):
+            if aik_enc is not None:
+                (blob, key) = aik_enc
+            else:
+                logger.warning("Agent %s failed encrypting AIK", agent_id)
```
Before returning you should still send a response. E.g.:
```
web_util.echo_json_response(self, 400, f"Error: failed encrypting AK")
```

mdrocco replied:
668a1bd

THS-on commented on keylime/tenant.py:
```diff
@@ -953,7 +959,7 @@ def do_quote(self):
 
             except Exception as e:
-                if response.status_code in (503, 504):
+                if response is not None and response.status_code in (503, 504):
```
```suggestion
                if response is None or response.status_code in (503, 504):
```
If we don't get anything back than also retry

mdrocco replied:
thanks!

THS-on commented on keylime/tenant.py:
```diff
@@ -983,7 +989,7 @@ def do_quote(self):
 
         if "results" not in response_body:
-            raise UserError(f"Error: unexpected http response body from Cloud Agent: {str(response.status)}")
+            raise UserError(f"Error: unexpected http response body from Cloud Agent: {str(response_body)}")
```
I don't know if we actually want to log the entire body. I think the status code is fine.

mdrocco replied:
da2288aded96fcf8ec3a9b3ece9dce994e173e55

THS-on commented on keylime/tpm/tpm_main.py:
```diff
@@ -1059,6 +1111,10 @@ def __tpm2_checkquote(self, pubaik, nonce, quoteFile, sigFile, pcrFile, hash_alg
                 "-q",
                 nonce,
             ]
+        else:
```
This case should never happen. Calling sys.exit() is fine for the functions only used by the agent, but for the verifier i think it is not a good idea. Maybe instead we can only add the check for the old version and remove the check for the newer ones.

mdrocco replied:
do you mean something like that? 71b33a9
If it looks good to you, I will apply the same change to all the spots where I added that "internal error" thing.

THS-on replied:
Yeah, seems the better option.

mdrocco replied:
great, I will apply the change everywhere.

mdrocco replied:
done

THS-on commented on keylime/tpm/tpm_main.py:
```diff
@@ -1331,12 +1387,14 @@ def sim_extend(self, hashval_1, hashval_0=None, hash_alg=None):
-            codecs.decode(hashval_0, "hex_codec")
-            + codecs.decode(self.hashdigest(hashval_1.encode("utf-8"), hash_alg), "hex_codec"),
+        hdata = self.hashdigest(hashval_1.encode("utf-8"), hash_alg)
+        assert hdata is not None
```
What can go wrong here that the value might be None?

mdrocco replied:
https://github.com/mdrocco/keylime/blob/71b33a96d084321be7dcb0ef78ba4c3a1edf8a66/keylime/tpm/tpm_main.py#L1395

`codecs.decode` does not accept `None` for the `obj` parameter.

THS-on replied:
Yeah, but can `self.hashdigest(...)` return None?

mdrocco replied:
I don't understand your question, the type checker is making the point about `codecs.decode`.

THS-on replied:
Ah ok, then keep the asserts. On another note I think it is fine to just assign it to same variable instead of using a new one or does the type checker than complain?

mdrocco replied:
I am not sure what you mean by "assign it to the same variable", but as long as we are not re-typing the variable (which I second the checker if it complains about, as it makes the code unnecessarily obscure) I am happy to hear any simplification you have to propose.

THS-on reviewed (changes requested):
Thank you for this PR! I really going to improve the code quality.

mdrocco commented on keylime/backport_dataclasses.py:
```diff
@@ -236,7 +236,7 @@ class Field:
     def __init__(self, default, default_factory, init, repr, hash, compare,
                  metadata):
-        self.name = None
+        self.name = ""
```
done

mdrocco commented on keylime/agentstates.py:
```diff
@@ -18,6 +18,12 @@ class TPMClockInfo:
     restartcount: int
     safe: int
 
+    def __init__(self, clock, resetcount, restartcount, safe):
```
I reverted this change, I guess this "fix" is actually overridden by a better one somewhere else, as the type checker is not complaining if I remove it.

aplanas commented on keylime/crypto.py:
```diff
@@ -64,7 +64,8 @@ def rsa_sign(key, message, paddingt="internal"):
 
-    if paddingt == "default":
+    else:
+        assert paddingt == "default"
```
I guess is to make sure that paddingt is "internal" | "default". IMHO and `if`, `elif` and a `else` that thrown an exception communicates better.

mdrocco replied:
please take a look: https://github.com/keylime/keylime/pull/1139/commits/e6b2344b0d1ab9b4c0d6ccfbb8399e2fe5dccf5c

mdrocco commented on keylime/cli/options.py:
```diff
@@ -29,4 +29,4 @@ def get_opts_error(args):
     if args.allowlist_sig_key and not (args.allowlist_sig or args.allowlist_sig_url):
         return True, "--allowlist-sig-key must have either --allowlist-sig or --allowlist-sig-url"
-    return False, None
+    return False, ""
```
done

mdrocco commented on keylime/cloud_verifier_tornado.py:
```diff
@@ -1103,9 +1104,19 @@ async def process_agent(agent, new_operational_state, failure=Failure(Component.
-                agent["last_event_id"] = failure.highest_severity_event.event_id
+            if (
+                agent.get("severity_level") is None
+                or failure.highest_severity is not None
```
done

mdrocco commented on keylime/cloud_verifier_tornado.py:
```diff
@@ -1173,8 +1184,11 @@ async def process_agent(agent, new_operational_state, failure=Failure(Component.
                 # set up a call back to check again
-                cb = functools.partial(invoke_get_quote, agent, ima_policy, False, timeout=timeout)
+                def cb():
+                    asyncio.run(invoke_get_quote(agent, ima_policy, False, timeout=timeout))
```
it is matter of typing rather than style: `partial()` returns an object that is not a function.

mdrocco commented on keylime/cloud_verifier_tornado.py:
```diff
@@ -1333,6 +1353,7 @@ def main():
 
     def server_process(task_id):
         logger.info("Starting server of process %s", task_id)
+        assert isinstance(engine, Engine)
```
I am not sure how to do that, as `engine` is a global variable that gets assigned the return value of a function call. Feel free to suggest anything, I am not very familiar with type hints.

THS-on replied:
In line 42 you can write:

```
    engine: Engine = DBEngineManager().make_engine("cloud_verifier")
```

mdrocco replied:
that raise a type error:

```console
Expression of type "MockConnection" cannot be assigned to declared type "Engine"
  "MockConnection" is incompatible with "Engine"
```

mdrocco commented on keylime/api_version.py:
```diff
@@ -42,19 +42,25 @@ def normalize_version(v):
 def major(v):
-    return version.parse(str(v)).major
+    version_ = version.parse(str(v))
+    if isinstance(version_, version.Version):
```
regardless what `parse` does with respect to exceptions, the assertion looks simpler and safer to me: 53f6d15

mdrocco commented on keylime/api_version.py:
```diff
@@ -42,19 +42,25 @@ def normalize_version(v):
 
 def minor(v):
-    return version.parse(str(v)).minor
+    version_ = version.parse(str(v))
```
also targeted by 53f6d15

mdrocco commented on keylime/cloud_verifier_common.py:
```diff
@@ -322,7 +322,7 @@ def validate_ima_policy_data(agent_data):
     # Validate exlude list contains valid regular expressions
     is_valid, _, err_msg = validators.valid_exclude_list(lists.get("exclude"))
-    if not is_valid:
+    if not is_valid and err_msg is not None:
```
6ff10100c3938bc323bbbb06e2fc2b597e74442d

mdrocco commented on keylime/cloud_verifier_tornado.py:
```diff
@@ -1198,7 +1212,10 @@ async def process_agent(agent, new_operational_state, failure=Failure(Component.
+
+                def cb_quote_retry():
+                    asyncio.run(invoke_get_quote(agent, ima_policy, True, timeout=timeout))
+
```
Former point: since `invoke_get_quote` takes 4 arguments, partial application with 4 parameters is equivalent to the complete `run` call, to the best of my understanding.

Latter point: I am not getting your point, can you please explain what makes the change not equivalent?

mdrocco replied:
@aplanas are you referring to the behavior with respect to async calls? I am struggling with it since I noticed my change breaks the code badly (e.g. integration tests are failing), maybe you have some suggestion based on your better understanding of Python's async and similar?

mdrocco commented on keylime/config.py:
```diff
@@ -27,7 +27,7 @@ def convert(data):
         return data.decode()
     if isinstance(data, dict):
-        return dict(map(convert, data.items()))
+        return dict(iter(map(convert, data.items())))
```
Without that "fix", the type checker says:

```console
error: Argument of type "map[str | dict[Any, Any] | tuple[Unknown, ...] | list[Unknown] | Unknown]" cannot be assigned to parameter "__iterable" of type "Iterable[tuple[_KT@dict, _VT@dict]]" in function "__init__"
    "map[str | dict[Any, Any] | tuple[Unknown, ...] | list[Unknown] | Unknown]" is incompatible with "Iterable[tuple[Any, Any]]"                                                         TypeVar "_T_co@Iterable" is covariant
        Type "str | dict[Any, Any] | tuple[Unknown, ...] | list[Unknown] | Unknown" cannot be assigned to type "tuple[Any, Any]"
          "str" is incompatible with "tuple[Any, Any]" (reportGeneralTypeIssues)
```

I am not sure the "fix" is the best thing to do, but at least makes the checker happy and does not modify the behavior, to the best of my understanding.

THS-on commented:
@mdrocco can you do a clean rebase on master, this makes it easier to review?

mdrocco commented:
I merged master into the branch, let me know it that works.

mdrocco commented:
/packit retest-failed

THS-on commented on keylime/cloud_verifier_tornado.py:
```diff
@@ -572,13 +573,13 @@ def post(self):
-                    except SQLAlchemyError as e:
-                        logger.error("SQLAlchemy Error for agent ID %s: %s", agent_id, e)
-                        raise e
+                        # Write the agent to the database, attaching associated stored policy
```
Yes make it None by default.

mdrocco replied:
dd7b1c8

THS-on commented on keylime/cmd/ima_emulator_adapter.py:
```diff
@@ -32,6 +32,7 @@ def measure_list(file_path, position, ima_hash_alg, pcr_hash_alg, search_val=Non
 
             entry = ast.Entry(line, None, ima_hash_alg=ima_hash_alg, pcr_hash_alg=pcr_hash_alg)
 
+            assert entry.pcr_template_hash is not None
```
No we should just error in the constructor if one of the hashes returns None.

mdrocco replied:
bff9d1f

THS-on commented on pyproject.toml:
```diff
@@ -14,3 +14,14 @@ include = '\.pyi?$'
+ignore = [
+    "keylime/backport_dataclasses.py",
+    "keylime/keylime_agent.py",
+    "keylime/elchecking/",
```
The elchecking module should be verified in the long run, but it is fine to exclude it for this PR.

mdrocco commented:
@THS-on I rebased on master and squashed, as I screwed up with commit signatures, let me know if you have concerns with that. I will now move to debugging what I broke with those type "fixes".

mdrocco commented:
Thanks for checking, I changed the author.

mdrocco commented:
/packit retest-failed

stefanberger commented on keylime/ima/ast.py:
```diff
@@ -29,11 +29,15 @@
 def get_FF_HASH(hash_alg: Hash) -> bytes:
-    return codecs.decode(b"f" * (hash_alg.get_size() // 4), "hex")
+    hsize = hash_alg.get_size()
+    assert hsize is not None
```
What is driving this that you add an assert here? Why would the hsize possibly be None?

mdrocco replied:
https://github.com/mdrocco/keylime/blob/314dd2c8de2eeae5209b688903678f959f41ec3c/keylime/common/algorithms.py#L48

stefanberger commented:
My general suggestion would be to split this PR up into many small easy-to-digest patches and PRs. There are some things that could go in very easily (changes in crypto.py for example) and others may not be so easy to digest...

mdrocco commented:
Sure, let us split the PR. Can you please mark the not-so-easy-to-digest changes? This would also help isolating what is making CI tests fail. Thanks!

stefanberger commented on keylime/ima/ast.py:
```diff
@@ -370,7 +374,9 @@ def __init__(
         self._bytes = self.mode.bytes()
-        self.pcr_template_hash = self._pcr_hash_alg.hash(self._bytes)
+        pcr_hash_alg_ = self._pcr_hash_alg.hash(self._bytes)
+        assert pcr_hash_alg_ is not None
```
Can the hash() of something be None?

stefanberger commented:
It looks like there are many different types of changes bunched together in this PR. I would split it by individual types of changes.

stefanberger commented:
> This would also help isolating what is making CI tests fail

For that I would rebase on master.

mdrocco commented:
Thanks @stefanberger for the thorough review, I will rebase on master.

mdrocco commented:
closing as split into #1164 #1173 #1180
